sandboxes/heka/output/heka_s3_partition.lua

```diff
@@ -0,0 +1,281 @@
+local max_file_handles      = read_config("max_file_handles") or 1000
+local max_file_size         = read_config("max_file_size") or 1024 * 1024 * 500
+local max_file_age          = read_config("max_file_age") or 60 * 60
+local flush_on_shutdown     = read_config("flush_on_shutdown")
+
+--[[
+## Heka Protobuf Message S3 Output Partitioner
+
+Batches message data into Heka protobuf stream files based on the specified path
+dimensions and copies them to S3 when they reach the maximum size or maximum
+age.
+
+### Configuration
+
+#### Dimension Specification File
+This file contains a JSON specification for how the data should be partitioned
+into files and the location each file will reside at in S3. The JSON is shared
+with other tools so it is not directly included in the configuration.
+
+TODO: This specifification should be expanded (when necessary) to include
+- header values
+- field/array index support
+- patterns instead of exact matches
+
+##### Sample Specification File
+```json
+  {
+    "version": 1,
+    "dimensions": [
+      {"field_name": "submissionDate",
+        "allowed_values": {"min": "20140120", "max": "20140125"}},
+      {"field_name": "sourceName",
+        "allowed_values": "*"},
+      {"field_name": "sourceVersion",
+        "allowed_values": "*"},
+      {"field_name": "reason",
+        "allowed_values": ["idle-daily","saved-session"]},
+      {"field_name": "appName",
+        "allowed_values": ["Firefox", "Fennec", "Thunderbird", "FirefoxOS", "B2G"]},
+      {"field_name": "appUpdateChannel",
+        "allowed_values": ["default", "nightly", "aurora", "beta", "release", "esr"]},
+      {"field_name": "appVersion",
+        "allowed_values": "*"}
+    ]
+  }
+```
+
+#### Sample Configuration
+
+```lua
+filename = "heka_s3_partition.lua"
+
+-- see the specification above
+dimension_file  = "foobar.json"
+
+-- directory location to store the intermediate output files
+batch_path       = "/var/tmp/foobar"
+
+-- Specifies how many data files to keep open at once. If there are more
+-- "current" files than this, the least-recently used file will be closed
+-- and then re-opened if more messages arrive before it is copied to S3. The
+-- default is 1000. A value of 0 means no maximum.
+max_file_handles    = 1000
+
+-- Specifies how much data (in bytes) can be written to a single file before
+-- it is copied to s3 (default 500MB)
+max_file_size       = 1024 * 1024 * 500
+
+-- Specifies how long (in seconds) to wait before it is copied to s3
+-- (default 1 hour).  Idle files are only checked every tickec_interval seconds.
+max_file_age        = 60 * 60
+
+-- Specifies that all local files will be copied S3 before exiting (default false).
+flush_on_shutdown = true
+preserve_data       = not flush_on_shutdown -- should always be the inverse of flush_on_shutdown
+
+s3_path             = "s3://foo"
+
+ticker_interval     = 60
+```
+--]]
+
+local function sanitize_dimension(d)
+    if type(d) == "string" then
```

Comment:
Also need to support number values (`sampleId`, for example) for dimensions. In the past we've used boolean fields as dimensions too (`telemetryEnabled`), so that could be useful too. We may want to allow anything that can sanely be converted into a string.